### PR TITLE
Using pattern model for each label to create pattern matches. Fix bug

### PR DESCRIPTION
**Pull request by JingJZ160**
- Use pattern model for labels
- Merge pattern models from different labels together. For the same entity, the tokens are the same. All matches are grouped together and sorted (easier for unit test). The scores are also saved although I don't know how we are going to display the scores for multiple labels.

### CODE REVIEW

File: ar/__init__.py [hunk 1]
```diff
@@ -197,9 +197,25 @@ def generate_annotation_requests(dbsession, task_id: int,
         __text_list.append(row['text'])
 
     # Pattern decorations
+    __pattern_decor_for_all_labels = []
     if task.get_pattern_model():
-        __pattern_decor = task.get_pattern_model().predict(__text_list,
-                                                           fancy=True)
+        labels = task.get_labels()  # ["hotdog", "fake"]
```
Comment (edu-gp) [inline ar/__init__.py:202]: Remove comment please

File: ar/__init__.py [hunk 2]
```diff
@@ -197,9 +197,25 @@ def generate_annotation_requests(dbsession, task_id: int,
...
+        logging.error(labels)
```
Comment (edu-gp) [inline ar/__init__.py:203]: Remove debug message

File: ar/__init__.py [hunk 3]
```diff
@@ -197,9 +197,25 @@ def generate_annotation_requests(dbsession, task_id: int,
         __text_list.append(row['text'])
 
     # Pattern decorations
+    __pattern_decor_for_all_labels = []
     if task.get_pattern_model():
```
Comment (edu-gp) [inline ar/__init__.py:201]: `task.get_pattern_model()` is no longer relevant...

File: ar/__init__.py [hunk 4]
```diff
@@ -197,9 +197,25 @@ def generate_annotation_requests(dbsession, task_id: int,
...
-        __pattern_decor = task.get_pattern_model().predict(__text_list,
-                                                           fancy=True)
+        labels = task.get_labels()  # ["hotdog", "fake"]
+        logging.error(labels)
+        for label in labels:
+            pattern_model_for_label = get_pattern_model_for_label(
+                dbsession=dbsession, label=label)
+            logging.error(pattern_model_for_label)
```
Comment (edu-gp) [inline ar/__init__.py:207]: Remove debug message

File: ar/__init__.py [hunk 5]
```diff
@@ -223,6 +239,49 @@ def generate_annotation_requests(dbsession, task_id: int,
     return annotation_requests
 
 
+def _merge_pattern_decor_for_all_labels_recursive(
+        pattern_decor_for_all_labels: List[List], low: int, high: int):
```
Comment (edu-gp) [inline ar/__init__.py:243]: Can you add some documentation?

And sorry can you also explain to me why we need to do (what looks to me like) a merge sort here?

File: ar/__init__.py (hunk 1 above)
Comment (JingJZ160) [reply ar/__init__.py:202]: Done.

File: ar/__init__.py (hunk 2 above)
Comment (JingJZ160) [reply ar/__init__.py:203]: Done.

File: ar/__init__.py (hunk 3 above)
Comment (JingJZ160) [reply ar/__init__.py:201]: Done.

File: ar/__init__.py (hunk 4 above)
Comment (JingJZ160) [reply ar/__init__.py:207]: Done.

File: ar/__init__.py (hunk 5 above)
Comment (JingJZ160) [reply ar/__init__.py:243]: As discussed, we are going to keep this logic. The documentation is added to the code.

**Comment by JingJZ160** [comment]
Local test was done by adding a fake pattern model for label "fake". The patterns includes "the" and "cloud". The other label is "hotdog" but the patterns are from the healthcare patterns.

Just to show that it works:
image

image

**Pull request merged**